Make Hook::unregisterHook able to unregistered many hooks at once

MDWeb-Lille opened the pull request:
Apply the same behavior as on Hook::registerHook

<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop 
| Description?  | Ability to pass an array in the Hook function unregisterHook as in the Hook method registerHook
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | 
| How to test?  | Put an array of hook name and hook id and/of only a hook name or id<br>$this->unregisterHook(array('header',12',validateOrder')); or $this->unregisterHook('header');

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/13587)
<!-- Reviewable:end -->

MDWeb-Lille committed:
Make Hook::unregisterHook able to unregistered many hooks at once

Apply the same behavior as on Hook::registerHook

MDWeb-Lille commented:
ping @mickaelandrieu

matks commented:
Hi ! I'm wondering; what is the benefit of doing
```
$this->unregisterHook($myModule, array('header',12',validateOrder'));
```
over
```
$this->unregisterHook($myModule, array(12));
$this->unregisterHook($myModule, array('header'));
$this->unregisterHook($myModule, array('validateOrder'));
```

Also, if we do that, maybe we should rather create a new function `public static function unregisterHooks($module_instance, $hook_names, $shop_list = null)` ?
Because we might optimize a bit things here :) for example you can delete all SQL entries with a single query.

PierreRambaud commented:
@matks It's to be consistent with what we are already able to do with `registerHook`  (https://github.com/PrestaShop/ps_facetedsearch/blob/dev/ps_facetedsearch.php#L150) but not with unregister: https://github.com/PrestaShop/ps_facetedsearch/commit/e5b293acaade81a17696c93188822287934642fc#diff-3327be713372041d22f3447af6020f94R61

i'm in favor of doing this. But yes, SQL can be optimized a little more

MDWeb-Lille committed:
Update SQL Delete Query

MDWeb-Lille commented:
I cleaned up the SQL a bit and moved the check on the `$shop_list` variable to do it once at the top of the loop and add it at the end the query.

I prefer not to delete all hooks in a single SQL query to keep the query between each hook call `actionModuleUnRegisterHookBefore` and `actionModuleUnRegisterHookAfter`

MDWeb-Lille committed:
CS Fixer - Fix space after comma

Quetzacoalt91 reviewed (dismissed):
In other circumstances, I would have asked you to add another method for handling several hook in a row.
But as we aim to be like the `registerHook` method, let's go for this.

sarahdib commented:
@MDWeb-Lille 
Can you provide me a module with this modification in order to test the PR.
Thank you for your contribution.

MDWeb-Lille commented:
Sorry for the delay !
If you can install and uninstall the module, it's OK!

[developer_tools.zip](https://github.com/PrestaShop/PrestaShop/files/3147698/developer_tools.zip)

Matt75 commented:
You don't have to unregisterHook on uninstall a Module, totally unnecessary : https://github.com/PrestaShop/PrestaShop/blob/799ef8416ced48274459fb241f340fba59de9849/classes/module/Module.php#L707

kpodemski commented:
@Matt75 you're absolutely right, but it makes sense if you want to unregister hook on module update or something :)

Matt75 commented:
Right, but you need to unregisterExceptions on this hook too on this case ;)
https://github.com/PrestaShop/PrestaShop/blob/799ef8416ced48274459fb241f340fba59de9849/classes/module/Module.php#L708

sarahdib commented:
@MDWeb-Lille Thank you for the contribution :)

Matt75 commented:
Not in favor of this change, modules developers forgot already to Module::unregisterExceptions() when unregister a hook.

This can enforce the misunderstanding about this mechanism. Table in database are already messed up with module developers not aware of that. This is mostly a lack in documentation.

Shop are not only for developers, in production merchants use exceptions mechanism, some modules register them in addition to merchant. Old shop who migrate on 1.7 have a lot of exceptions because this mechanism was useful in the past, now it's forgotten.

There are other way to handle Module::unregisterHook() more efficiency. For example :
```php
class homedisplayexample extends Module
{
    /**
     * @var array List of hooks you use
     */
    public $hooks = array(
        'displayHome', // For example
    );

    /**
     * Constructor
     */
    public function __construct()
    {
        $this->name = 'homedisplayexample';
        $this->tab = 'administration';
        $this->version = '1.0.0';
        $this->author = 'Matt75';
        $this->need_instance = 0;

        parent::__construct();

        $this->displayName = $this->l('Home display example');
        $this->description = $this->l('Example of module to display on homepage');
    }

    /**
     * Install Module
     *
     * @return bool
     */
    public function install()
    {
        return parent::install()
            && $this->registerHook($this->hooks) // Allow to use an array of hook names
        ;
    }

    /**
     * Uninstall Module
     *
     * @return bool
     */
    public function uninstall()
    {
        return parent::uninstall(); // Just to demonstrate no need to unregisterHook !
    }

    /**
     * Retrieve hooks used by the module
     *
     * @return array
     */
    protected function getHooksUsed()
    {
        $query = new DbQuery();
        $query->select('h.name');
        $query->from('hook', 'h');
        $query->innerJoin('hook_module', 'hm', 'hm.id_hook = h.id_hook');
        $query->where('hm.id_module = ' . (int)$this->id);
        $query->groupBy('h.name');

        $hooks = Db::getInstance()->executeS($query);

        if (!$hooks) {
            return [];
        }

        return array_column($hooks, 'name');
    }

    /**
     * Clear hooks used by the module
     * This method is public because it used in upgrade file
     *
     * @return bool
     */
    public function clearHookUsed()
    {
        $result = true;
        $hooksUsed = $this->getHooksUsed();

        foreach ($this->hooks as $hook) {
            if (!in_array($hook, $hooksUsed)) {
                $result &= $this->registerHook($hook);
            }
        }

        foreach ($hooksUsed as $hookUsed) {
            if (!in_array($hookUsed, $this->hooks)) {
                $result &= $this->unregisterHook($hookUsed);
                $result &= $this->unregisterExceptions($hookUsed);
            }
        }

        return $result;
    }

    /**
     * Display something on homepage
     *
     * @param array $params
     *
     * @return string
     */
    public function hookDisplayHome($params)
    {
        $this->smarty->assign(array(
            'hook_name' => 'hookDisplayHome', // For example
        ));

        return $this->display(__FILE__, 'displayHome.tpl');
    }
﻿}
```

We can improve documentation or add a new helper function to handle this in the core but how do you handle retro compatibility ?

Module developers sell on Addons can’t use this change because it will work only for 1.7.6 so I think it’s better to improve documentation and improve modules in @friends-of-prestashop

Not in favor to make improvement in Module legacy class too : https://github.com/PrestaShop/PrestaShop/pull/13201#issuecomment-480585602

Matt75 commented:
> We can improve documentation or add a new helper function to handle this in the core but how do you handle retro compatibility ?

Add a new helper to handle that properly in src folder, in an Adapter, existing one.
Usage for developer who doesn’t want to rely on legacy.
There is a better way to explore

Matt75 reviewed (changes requested):
Do not make improvements in legacy classes, please handle Module::unregisterExceptions() too

mickaelandrieu commented:
Hi @Mat75,

I wasn't aware of the exceptions linked to the hooks (and why do we need that at the first place? :thinking: ).

Should this contribution be closed or improved? And if an improvement is expected, you need to explain more what you expect

> Do not make improvements in legacy classes, please handle Module::unregisterExceptions() too

For instance, this comment is not really understandable to me :/

Mickaël

Matt75 commented:
@mickaelandrieu We need to unregisterExceptions in order to remove entries in associated table before unregisterHook unless we lost the associated identifier so we cannot clear exception table as I remember. I’m on my phone on vacation so I cannot check that now.

I think this can be improved and not closed.

For my latest comment, please read my comment in the other issue linked, it refer to article about architecture on build blog

PierreRambaud commented:
Because you're currently talking, I add the label "Waiting for dev". 
Ping @matks Could be interresting to have your point of view here!

matks commented:
Ok, I'm not sure I understand everything here.

From what I understand:
- PR enables to unregister an array of hooks instead of a single one, and there's nothing wrong with that
- @Matt75 states that the function `unregisterHook()` is often used in a bad way by module developers which use it in `uninstall module` process but forget to unregister Exceptions too
- @Matt75 states that the standard `Module::uninstall()` function already performs the required hooks unregistration and exception unregistration steps, so in a module it's better to let PrestaShop do this job (rather than unregistering things by yourself)

Is it right ?

So it seems to me it's 2 different topics 🤔?
I suggest we
- accept and merge the PR because it can be useful (as @kpodemski suggested in a `update module` process for example)
- but also work on better communication towards module developers about "what should happen in a good uninstall process"

We could also try to detect forgotten exceptions and hooks after the uninstall and warn the merchant.

Matt75 commented:
@matks I think merging this like that will increase the misunderstanding about how it should work really.

As I say, this is useless because it's already handle by core on Module::uninstall() and for upgrade script, we can do better.

Before doing a Module::unregisterHook() we ALWAYS do a Module::unregisterException() without the exception table will be inconsistent with orphan entries.

Problem here is lack of documentation of historical PrestaShop process.

Hook Exception is used by merchants for some use case, less than past but still used

Matt75 commented:
What about do not make improvements in legacy classes ?

For good improvement why not, but here I'm absolutely not convinced.

I think it's not a good idea because module developers do already many bad things when using current implementation of Module::unregisterHook()

Matt75 commented:
By the way, who will use this ?

As an Addons Developers, I need to make module work with much possible versions of PrestaShop.

So if I want to use this new possibility proposed here, I need to manage retro compatibility right ?

So with this change or without, I have to write code to handle it in all versions supported by the module.

So where is value here ?

Sinepel commented:
Hello everyone,

as a Prestashop developer in an agency, this function is very useful and this refactoring is for me, necessary.

During our specific developments for our customers, we regularly use this function and this modification can shorten the code of a few useful lines.

In addition, Prestashop developer for over 5 years now, I know few developers who know that hooks are unregistered when uninstalling a module.

Matt75 commented:
So as an agency you want coding change in the core but you need to handle it by your side for retro compatibility with PrestaShop version previous 1.7.7 ?

I was PrestaShop developer in agency during 6 years, I have to handle retro compatibility of modules selled on Addons for PrestaShop 1.4 to 1.6 and last few months to 1.7

I need to make additional code to handle new functionalities and coding changes made by PrestaShop in module for retro compatibility.

The better strategy for me was to develop external class responsible to handle common task like unregisterHook with unregisterException and integrate it on every modules I made. In fact we created a library like the old Backward Module. https://addons.prestashop.com/en/data-migration-backup/6222-.html
But in a better way of course.

For me this change need to be done in custom Adapter or in Core Adapters but not in legacy classes.

Did you know about Hook Exception and need to use unregisterExceptions ? Do you handle it on your side before ?

I think you are on the wrong way, give me more real arguments please.

PierreRambaud commented:
I think we can add new functionnality, not saying removing the old behavior, so for the 1.8, we will have a better solution 😸 
We need to improve code & behavior as much as possibile, but keep noticed since when it exists.

Matt75 commented:
So same change need to be done on unregisterExceptions to stay consistency. This should be documented and WARNING need to be announced to handle unregisterExceptions and about retro compatibility.
Sadly I’m convinced change behavior in legacy classes is a very bad way.
We have adapters for that and better strategy...

matks commented:
> Sadly I’m convinced change behavior in legacy classes is a very bad way.

I dont understand why you say we change behavior in Hook legacy class. Previous behavior is preserved, but we give more flexibility to users by allowing to pass an array of hooks instead of a single one. So this is new feature but previous behavior is kept unchanged.

Retro compatibility is about "do not break something which used to work", not "allow new features to work with previous version" 🤔. But indeed this will only be used by agency developers who control the version of the shop, modules developers wont be able to make use of it if they want to target a large compatibility range.

Also you seem very worried about possible bad uses of this function @Matt75, what do you have in mind ? I dont see what this change can make worse.

Matt75 commented:
Simple, you can take a look inside modules code...

Actually, we know module developers don’t do this process correctly, its not the most bad things they do, ok. It affect only one table who take care ? PrestaShop database is inconsistent and incoherent since the beginning.

But we have opportunity here to do something more better and we do the most simple and do not responsabilize module developers...

If we want increase quality of modules, we have to make better process and documentation. I know this is currently a work in progress.

I think a better way can be found here. That’s it

I was a module developers in 2 agencies in the past (4 years), 2 years as freelance on PrestaShop projects and I know what a module developer need.

They want guidelines and documentation because actually they need to lean by themselves and they cannot do take care how it work in details (lack of time). This is explained nowhere actually. Most of them are student in agencies or junior developers, **in my experience** (I talk only from my experience about what I lived before).

Work done currently on PrestaShop is really great and exciting, I learn all days by reading Github but most of modules developers cannot take this time.

Devdoc is not popular, most of developers doesn't read it because they know documentation is light. They are not watcher on Devdoc repository, they don't know we improve it since few months. Most of them still reading old documentation about 1.5 and 1.6 because we have a new website for current devdoc... I know most of them don't know about Devdoc, too old bookmarks on their browser... After read it one time and don't find answer about what they need, most of them forget it.

So they go on PrestaShop forums, they read some bad recommandations, try code change inside PrestaShop core or try an override... Long story...

That's why I think it's your responsibility to give them some good stuffs and I'm sorry but this change actually cannot be a good improvement because the problem about Hook Exceptions will be staying under carpet.

If this is only to not write 5 lines of code on their side and don't take care of Module::unregisterException() ... By the way, we can do better, for example I learn some cool native PHP functions some months ago :
```php
    /**
     * Uninstall Module
     *
     * No need to do this because it's already done by core in Module::uninstall()
     * @Sinepel I write only one line of code...
     *
     *
     * @return bool
     */
    public function uninstall()
    {
        return parent::uninstall()
        	&& array_product(array_map([$this, 'unregisterHook'], ['actionAdminOrdersListingFieldsModifier', 'actionAdminOrdersListingResultsModifier', 'actionAdminControllerSetMedia']));
    }
```
Maybe it's not so good, but it's funny to try.
```php
/**
 * This is an upgrade script for my previous Module class in this discussion
 * In the new version, I don't need anymore this hooks :
 *      - actionAdminOrdersListingFieldsModifier,
 *      - actionAdminOrdersListingResultsModifier
 *      - actionAdminControllerSetMedia
 * So I unregister them in one line...
 * No need to make core modifications...
 * And this work on every PrestaShop version
 * 
 * @param homedisplayexample $module Instance of my previous Module class in this discussion
 *
 * @return bool
 */
function upgrade_module_1_0_1($module)
{
    return array_product(array_map([$module, 'unregisterHook'], ['actionAdminOrdersListingFieldsModifier', 'actionAdminOrdersListingResultsModifier', 'actionAdminControllerSetMedia']));
}
```

I am exhausted, so sorry if I hurt someone, it's not what I have in mind.

matks commented:
After discussing further this matter with @Matt75 we agreed on a satisfying outcome:
- merge this PR with additionnal warning comments in order to highlight:
    -  the fact that this is only for 1.7.6 and that module developers who target an extended compatibility range should'nt use it
    - the fact that PrestaShop perform this in its `Module::uninstall()` function so it's useless to do it in this case
- update `unregisterExceptions()` to be consistent with this change, same work to do with warnings
- create reference modules to be used as examples to educate module developers about the things to be wary of (see https://github.com/PrestaShop/PrestaShop/issues/14005)

Progi1984 commented:
Hello @MDWeb-Lille, thanks for your contribution, would you mind to rebase your pull request? Thanks

Commit:
Fix cs

Progi1984 commented:
> Hello @Progi1984 It's rebased !

@MDWeb-Lille Great. Could you fix PHPCSFixer ?
In your local environment, execute `./vendor/bin/php-cs-fixer fix --dry-run --diff` ;) and fix errors :)

MDWeb-Lille commented:
@PierreRambaud was faster !

Matt75 commented:
Phpdoc of this method should be updated and PrestaShop version availability for this change should be mentioned to help developers. (I still find this change stupid ^^)

PierreRambaud commented:
PR rebased :)

eternoendless commented:
I'm sorry to get to this so late in the discussion, but I honestly struggle to see the value added by this change.

The exact same effect can be achieved by doing this:

```php
$hooks = ['someHook', 'someOtherHook];
foreach ($hooks as $hook) {
	$this->unregisterHook($hook);
}
```

There's basically no performance gain in allowing to pass along an array to `unregisterHook`. We're basically talking about a micro optimization that further complexifies existing code by adding mixed type parameters.

I'm sorry but I'm against this change.

PierreRambaud commented:
> I'm sorry to get to this so late in the discussion, but I honestly struggle to see the value added by this change.
> 
> The exact same effect can be achieved by doing this:
> 
> ```
> $hooks = ['someHook', 'someOtherHook];
> foreach ($hooks as $hook) {
> 	$this->unregisterHook($hook);
> }
> ```
> 
> There's basically no performance gain in allowing to pass along an array to `unregisterHook`. We're basically talking about a micro optimization that further complexifies existing code by adding mixed type parameters.
> 
> I'm sorry but I'm against this change.

We must be consistent with the `registerHook`, if we refuse this pull request, we should mark as deprecated the capability to pass an array in `registerHook` method. 
This kind of change has clearly no performance effect, it's more about simplicity and readability.

matks commented:
Hello @Sinepel it seems I was never able to find some time to finish this PR 😞  and I don't think I'll find it in the future. Should I close it or do you wish to finish it?

PierreRambaud commented:
Hi @MDWeb-Lille ,

Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant. Otherwise, maybe someone from the @PrestaShop/prestashop-core-developers will be able to continue the work you've already done!

Thanks!
